execute-generators/src/main/kotlin/de/itemis/mps/gradle/generate/Helper.kt

```diff
@@ -0,0 +1,110 @@
+package de.itemis.mps.gradle.generate
+
+import com.intellij.openapi.util.AsyncResult
+import jetbrains.mps.make.MakeSession
+import jetbrains.mps.make.facet.FacetRegistry
+import jetbrains.mps.make.facet.IFacet
+import jetbrains.mps.make.facet.ITarget
+import jetbrains.mps.make.script.IScript
+import jetbrains.mps.make.script.ScriptBuilder
+import jetbrains.mps.messages.IMessage
+import jetbrains.mps.messages.IMessageHandler
+import jetbrains.mps.messages.MessageKind
+import jetbrains.mps.project.Project
+import jetbrains.mps.smodel.language.LanguageRegistry
+import jetbrains.mps.smodel.resources.ModelsToResources
+import jetbrains.mps.smodel.runtime.MakeAspectDescriptor
+import jetbrains.mps.tool.builder.make.BuildMakeService
+import org.apache.log4j.Logger
+import org.jetbrains.mps.openapi.language.SLanguage
+import org.jetbrains.mps.openapi.model.SModel
+
+private val logger = Logger.getLogger("de.itemis.mps.gradle.generate")
+
+private val DEFAULT_FACETS = listOf(
+        IFacet.Name("jetbrains.mps.lang.core.Generate"),
+        IFacet.Name("jetbrains.mps.lang.core.TextGen"),
+        IFacet.Name("jetbrains.mps.make.facets.Make"))
+
+private class MsgHandler : IMessageHandler {
+    val logger = Logger.getLogger("de.itemis.mps.gradle.generate.messages")
+    override fun handle(msg: IMessage) {
+        when (msg.kind) {
+            MessageKind.INFORMATION -> logger.info(msg.text, msg.exception)
+            MessageKind.WARNING -> logger.warn(msg.text, msg.exception)
+            MessageKind.ERROR -> logger.error(msg.text, msg.exception)
+        }
+    }
+
+}
+
+private fun createScript(proj: Project, models: List<org.jetbrains.mps.openapi.model.SModel>): IScript {
+
+    val allUsedLanguagesAR: AsyncResult<List<SLanguage>> = AsyncResult()
+
+    proj.modelAccess.runReadAction {
+        allUsedLanguagesAR.setDone(models.map { it.module }.distinct().flatMap { it.usedLanguages }.distinct())
```

Comment:
We should also take into account the languages that the used languages extend

Reply:
If not sure but you might be correct. It's pretty hard to track what MPS is doing in the `ModulesClusterizer` and friends. But seems like `ModulesCluster.fillEdges` is in fact doing a ` new SLanguageHierarchy(myLanguageRegistry, moduleUsedLanguages).getExtended()`. Will change the implementation.

Reply:
Done.